src/openrct2-ui/windows/NetworkStatus.cpp

```diff
@@ -31,135 +29,139 @@ static rct_widget window_network_status_widgets[] = {
+    void OnMouseUp(WidgetIndex widgetIndex) override
+    {
+        switch (widgetIndex)
+        {
+            case WIDX_CLOSE:
+                Close();
+                break;
+        }
+    }
 
-    return window;
-}
+    void OnUpdate() override
+    {
+        InvalidateWidget(WIDX_BACKGROUND);
+    }
 
-static void WindowNetworkStatusOnclose(rct_window* w)
-{
-    if (_onClose != nullptr)
+    void OnTextInput(WidgetIndex widgetIndex, std::string_view text) override
     {
-        _onClose();
+        _password.clear();
+        switch (widgetIndex)
+        {
+            case WIDX_PASSWORD:
+                _password = text;
+                break;
+        }
+        if (text.empty())
+        {
+            network_shutdown_client();
+        }
+        else
+        {
+            network_send_password(_password);
+        }
     }
-}
 
-static void WindowNetworkStatusMouseup(rct_window* w, WidgetIndex widgetIndex)
-{
-    switch (widgetIndex)
+    void OnPrepareDraw() override
     {
-        case WIDX_CLOSE:
-            window_close(*w);
-            break;
+        widgets[WIDX_BACKGROUND].right = width - 1;
+        widgets[WIDX_BACKGROUND].bottom = height - 1;
+        widgets[WIDX_TITLE].right = width - 2;
+        widgets[WIDX_CLOSE].left = width - 2 - 0x0B;
+        widgets[WIDX_CLOSE].right = width - 2 - 0x0B + 0x0A;
     }
-}
 
-static void WindowNetworkStatusUpdate(rct_window* w)
-{
-    widget_invalidate(*w, WIDX_BACKGROUND);
-}
+    void OnDraw(rct_drawpixelinfo& dpi) override
+    {
+        WindowDrawWidgets(*this, &dpi);
+        thread_local std::string _buffer;
+        _buffer.assign("{BLACK}");
+        _buffer += _windowNetworkStatusText;
+        gfx_clip_string(_buffer.data(), widgets[WIDX_BACKGROUND].right - 50, FontSpriteBase::MEDIUM);
+        ScreenCoordsXY screenCoords(windowPos.x + (width / 2), windowPos.y + (height / 2));
+        screenCoords.x -= gfx_get_string_width(_buffer, FontSpriteBase::MEDIUM) / 2;
+        gfx_draw_string(&dpi, screenCoords, _buffer.c_str());
+    }
 
-static void WindowNetworkStatusTextinput(rct_window* w, WidgetIndex widgetIndex, char* text)
-{
-    _password[0] = '\0';
-    switch (widgetIndex)
+    void SetCloseCallBack(close_callback onClose)
     {
-        case WIDX_PASSWORD:
-            if (text != nullptr)
-                safe_strcpy(_password, text, sizeof(_password));
-            break;
+        _onClose = onClose;
     }
-    if (text == nullptr)
+
+    void SetWindowNetworkStatusText(const char* text)
     {
-        network_shutdown_client();
+        safe_strcpy(_windowNetworkStatusText, text, sizeof(_windowNetworkStatusText));
     }
-    else
+
+    void SetPassword(char* password)
     {
-        network_send_password(_password);
+        _password = password;
     }
+
+private:
+    close_callback _onClose = nullptr;
+    char _windowNetworkStatusText[1024];
```

Comment:
This could be changed to a `std::string` as well.

Reply:
Done. Included an upper change in WindowManager to avoid converting to const char*.